### PR TITLE
Fixing broken hierarchy for items list in Combobox

### LINKED ISSUE #3654: Accessibility: The list item of comboBox incorrectly displayed in inspect tree when setting its DropDownStyle as DropDown or DropDownList.
* .NET Core Version: 
.NET SDK 5.0.100-rc.1.20377.3

* More Info:
This is not a regression issue, it also reproduce on .NET Core 3.1.

* Have you experienced this same bug with .NET Framework?: 
No

**Problem description:**
When setting ComboBox's DropDownStyle as DropDown or DropDownList, the list item of comboBox doesn't display below Form1 window node, it incorrectly displayed below Desktop1 panel node.
Core

**Expected behavior:**
The list items of comboBox's UIA hierarchy should display below Form1 window node.
Framework

**Minimal repro:**
1. Launch Inspect tool.
2. Create WinForms core project with a comboBox control.
3. Set the DropDownStyle property of comboBox as DropDown or DropDownList.
4. Run the project.
5. Focus on the list items of comboBox, then observe the UIA hierarchy in inspect tree.

### PR DESCRIPTION
**Pull request by SergeySmirnov-Akvelon**
Fixes #3654

## Proposed changes

- Updated logic of "FragmentNavigate" method for Combobox accessible objects. 
- Updated logic for ComboBox.ComboBoxChildNativeWindow. Now we return our ComboBoxChildEditUiaProvider instead of the native.
- Fixed unit tests

## Customer Impact
Before:
Issue-3654-broken
After:
Issue-3654-fixed

## Regression? 

- Yes

## Risk

- Minimal

## Test methodology 

- Manually 
- CTI team

## Accessibility testing  
- Inspect
- Accessibility Insights
- Narrator 

## Test environment(s) 
- Microsoft Windows [Version 10.0.19041.388]
- .Net SDK 5.0.100-rc.2.20479.15

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/4185)

### CODE REVIEW

**Comment by SergeySmirnov-Akvelon** [comment]
CTI approved

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxAccessibleObject.cs [hunk 1]
```diff
@@ -333,6 +330,26 @@ internal override void SetFocus()
...
+            {
+                if (_owningComboBox.DroppedDown)
+                {
+                    return _owningComboBox.ChildListAccessibleObject;
+                }
+
+                switch (_owningComboBox.DropDownStyle)
+                {
+                    case ComboBoxStyle.DropDown:
+                        return _owningComboBox.IsHandleCreated ? _owningComboBox.ChildEditAccessibleObject : DropDownButtonUiaProvider;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxAccessibleObject.cs:344]: please re-format this line because it's longer than 120 chars

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs [hunk 2]
```diff
@@ -88,16 +88,11 @@ internal override UiaCore.IRawElementProviderFragment FragmentNavigate(UiaCore.N
...
-                        int comboBoxChildFragmentCount = comboBoxAccessibleObject.GetChildFragmentCount();
-                        if (comboBoxChildFragmentCount > 1)
-                        { // Text or edit is previous;
-                            return comboBoxAccessibleObject.GetChildFragment(comboBoxChildFragmentCount - 1);
-                        }
-                    }
-
-                    return null;
+                    return _owner.DropDownStyle == ComboBoxStyle.DropDownList
+                            ? _owner.ChildTextAccessibleObject
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs:92]: Why are you not testing is handle had been create for DropDownList style? 
I see there are more places where this question would apply in this file, could you please explain when we have to test if handle was created?

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good. Please clarify, why we are not testing if Handle was created everywhere before creating providers or accessible object?
@vladimir-krestov - Had you reviewed this change?

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs [hunk 3]
```diff
@@ -78,16 +78,9 @@ private ChildAccessibleObject GetChildAccessibleObject(ChildWindowType childWind
...
+            internal void WmGetObject(ref Message m)
             {
-                if (m.LParam == (IntPtr)NativeMethods.UiaRootObjectId &&
-                    // Managed UIAutomation providers are supplied for child list windows but not for the child edit window.
-                    // Child list accessibility object provides all necessary patterns and UIAutomation notifications,
-                    // so there is no need to native provider supplement.
-                    // Child edit accessibility object has only partial support of edit box accessibility, most of the patterns
-                    // and notifications for child edit window are supplied by native providers, so here is no need to
-                    // override root UIA object for child edit window.
-                    (_childWindowType == ChildWindowType.ListBox || _childWindowType == ChildWindowType.DropDownList))
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs:90]: Could you please explain why did you remove this additional check?

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ChildAccessibleObjectTests.cs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+using static Interop;
+
+namespace System.Windows.Forms.Tests
+{
+    public class ComboBox_ChildAccessibleObjectTests : IClassFixture<ThreadExceptionFixture>
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ChildAccessibleObjectTests.cs:11]: The class name doesn't consider tested types. I think you need to divide tests for `ComboBox` children as separate files (`ComboBox_ComboBoxChildTextUiaProvider`, `ComboBox_ComboBoxChildListUiaProvider`, etc.) and test them in  corresponding places.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ChildAccessibleObjectTests.cs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        [InlineData(ComboBoxStyle.DropDown)]
+        public void DropDownButtonUiaProvider_FragmentNavigate_PreviousSibling_ReturnsExpected_IfHanleIsNotCreated(ComboBoxStyle comboBoxStyle)
+        {
+            using ComboBox comboBox = new ComboBox
+            {
+                DropDownStyle = comboBoxStyle
+            };
+
+            AccessibleObject previousItem = GetComboBoxAccessibleObject(comboBox).DropDownButtonUiaProvider.FragmentNavigate(UiaCore.NavigateDirection.PreviousSibling) as AccessibleObject;
+            var expectedItem = comboBoxStyle == ComboBoxStyle.DropDownList ? comboBox.ChildTextAccessibleObject : null;
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ChildAccessibleObjectTests.cs:203]: Make the type explicit (remove `var`)

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs [hunk 6]
```diff
@@ -78,16 +78,9 @@ private ChildAccessibleObject GetChildAccessibleObject(ChildWindowType childWind
...
+               if (m.LParam == (IntPtr)NativeMethods.UiaRootObjectId)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs:83]: Add one space before `if` (just use auto refactoring)

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs [hunk 7]
```diff
@@ -74,28 +75,49 @@ public void ComboBoxAccessibleObject_ExpandCollapse_Set_CollapsedState_IfControl
...
         {
-            using var control = new ComboBox
+            using var comboBox = new ComboBox
             {
-                DropDownStyle = comboBoxStyle
+                DropDownStyle = ComboBoxStyle.DropDown
             };
 
-            if (createControl)
+            var accessibleObject = comboBox.AccessibilityObject;
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs:86]: Change `var` everywhere

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs [hunk 8]
```diff
@@ -74,28 +75,49 @@ public void ComboBoxAccessibleObject_ExpandCollapse_Set_CollapsedState_IfControl
...
         }
 
-        [WinFormsTheory]
-        [InlineData(ComboBoxStyle.DropDown, true)]
-        [InlineData(ComboBoxStyle.DropDownList, true)]
-        [InlineData(ComboBoxStyle.DropDown, false)]
-        [InlineData(ComboBoxStyle.DropDownList, false)]
-        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_NotNull(ComboBoxStyle comboBoxStyle, bool createControl)
+        [WinFormsFact]
+        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_DropDown_ReturnExpected_IfHandleNotCreated()
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs:79]: Don't forget to change the test name when you change the `FragmentNavigate` logic

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs [hunk 9]
```diff
@@ -74,28 +75,49 @@ public void ComboBoxAccessibleObject_ExpandCollapse_Set_CollapsedState_IfControl
...
-            if (createControl)
+            var accessibleObject = comboBox.AccessibilityObject;
+            UiaCore.IRawElementProviderFragment firstChild = accessibleObject.FragmentNavigate(UiaCore.NavigateDirection.FirstChild);
+
+            Assert.Equal(GetComboBoxAccessibleObject(comboBox).DropDownButtonUiaProvider, firstChild);
+            Assert.False(comboBox.IsHandleCreated);
+        }
+
+        [WinFormsFact]
+        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_DropDownList_ReturnExpected_IfHandleNotCreated()
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs:94]: ditto

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs [hunk 10]
```diff
@@ -109,5 +131,116 @@ public void ComboBoxEditAccessibleObject_NameNotNull(string name)
...
+                        bool childListDisplayed = droppedDown || comboBoxStyle == ComboBoxStyle.Simple;
+                        yield return new object[] { comboBoxStyle, createControl, droppedDown, childListDisplayed };
+                    }
+                }
+            }
+        }
+
+        [WinFormsTheory]
+        [MemberData(nameof(ComboBoxAccessibleObject_FragmentNavigate_FirstChild_ReturnsExpected_TestData))]
+        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_ReturnsExpected(ComboBoxStyle comboBoxStyle, bool createControl, bool droppedDown, bool childListDisplayed)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs:152]: Move arguments to a new line

**Review by vladimir-krestov — COMMENTED** [review]
Good work! Added cosmetic review points

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxAccessibleObject.cs [hunk 11]
```diff
@@ -160,20 +160,17 @@ public ComboBoxChildDropDownButtonUiaProvider DropDownButtonUiaProvider
...
-                    if (childFragmentCount > 0)
-                    {
-                        return GetChildFragment(childFragmentCount - 1);
-                    }
+                    case UiaCore.NavigateDirection.FirstChild:
+                        return GetFirstChild();
+                    case UiaCore.NavigateDirection.LastChild:
+                        return _owningComboBox.DropDownStyle == ComboBoxStyle.Simple
+                            ? _owningComboBox.IsHandleCreated ? _owningComboBox.ChildEditAccessibleObject : null
+                            : DropDownButtonUiaProvider;
```
[the remark this answers is not in the document]
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxAccessibleObject.cs:170]: This was done to simulate the old behavior. On the other side, if the handle is not created, then the Accessibility tool will not be able to access such control. As result, I just added a check that if the handle is not created, return the null

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs (hunk 2 above)
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs:92]: This was done to simulate the old behavior. On the other side, if the handle is not created, then the Accessibility tool will not be able to access such control. As result, I just added a check that if the handle is not created, return the null

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs (hunk 3 above)
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildNativeWindow.cs:90]: This was done in order to return our accessibility object for edit control instead of our own accessibility object.

**Comment by SergeySmirnov-Akvelon** [comment]
Hi @Tanya-Solyanik. I verified manually ToolStripCombobox, DataGridViewComboBoxColumn and comboboxes from the test application including the data-bound combobox. We found no change in MSAA behavior because MCAA uses native controls and does not call the "FragmentNavigate" method. During the refactoring, we have made a lot of new changes, so I suggest to send a fix to CTI team once again after fixing review notes

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs (hunk 2 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs:95]: > if the handle is not created, then the Accessibility tool will not be able to access such control

This is not necessarily true. Suppose your combobox is shown but the drop down is not yet created and inspect focuses on the combo. Then I expand combobox node in inspect and inspect accesses the drop-down and forces drop down handle creation. In this sample scenario we want to create handle, but there there were other scenarios when it was a problem. But if this was copied from the old code, we can leave it as is.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox .ComboBoxChildListUiaProviderTests.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox .ComboBoxChildListUiaProviderTests.cs:1]: This file has the incorrect name: please remove the extra space after `ComboBox` in the file name

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs [hunk 13]
```diff
@@ -75,39 +76,136 @@ public void ComboBoxAccessibleObject_ExpandCollapse_Set_CollapsedState_IfControl
...
-        [InlineData(ComboBoxStyle.DropDown, false)]
-        [InlineData(ComboBoxStyle.DropDownList, false)]
-        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_NotNull(ComboBoxStyle comboBoxStyle, bool createControl)
+        [InlineData(ComboBoxStyle.DropDown)]
+        [InlineData(ComboBoxStyle.DropDownList)]
+        [InlineData(ComboBoxStyle.Simple)]
+        public void ComboBoxAccessibleObject_FragmentNavigate_FirstChild_ReturnNull_IfHandleNotCreated(ComboBoxStyle comboBoxStyle)
         {
-            using var control = new ComboBox
+            using var comboBox = new ComboBox
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ComboBox.ComboBoxAccessibleObjectTests.cs:84]: Please change `var` everywhere

**Review by vladimir-krestov — COMMENTED** [review]
Little points. In the rest, looks good to me

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs (hunk 2 above)
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildDropDownButtonUiaProvider.cs:95]: The dropdown list is not available in the Inspect and the Accessibility Insights when the dropdown list is not opened. This is the behavior of the native control and is also reproduced in .NET framework 4.7.2, .NET Core 3.1, .NET Core 5.0
Issue-3654-combobox

**Review by RussKie — APPROVED** [review]
:shipit:
